src/providers/krb5/krb5_child_handler.c

```diff
@@ -530,35 +596,64 @@ struct tevent_req *handle_child_send(TALLOC_CTX *mem_ctx,
         return NULL;
     }
 
+    if (kr->krb5_ctx->io_table == NULL) {
+        /* Create IO/pipe table if not exist. */
+        kr->krb5_ctx->io_table = sss_ptr_hash_create(kr->krb5_ctx, NULL, NULL);
+        if (kr->krb5_ctx->io_table == NULL) {
+            ret = ENOMEM;
+            goto fail;
+        }
+    }
+
     state->ev = ev;
     state->kr = kr;
     state->buf = NULL;
     state->len = 0;
     state->child_pid = -1;
     state->timeout_handler = NULL;
 
-    state->io = talloc(state, struct child_io_fds);
-    if (state->io == NULL) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "talloc failed.\n");
-        ret = ENOMEM;
-        goto fail;
-    }
-    state->io->write_to_child_fd = -1;
-    state->io->read_from_child_fd = -1;
-    talloc_set_destructor((void *) state->io, child_io_destructor);
-
     ret = create_send_buffer(kr, &buf);
     if (ret != EOK) {
         DEBUG(SSSDBG_CRIT_FAILURE, "create_send_buffer failed.\n");
         goto fail;
     }
 
-    ret = fork_child(req);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "fork_child failed.\n");
-        goto fail;
+    if (kr->pd->child_pid == 0) {
+        /* Create new child. */
+        ret = fork_child(ev, kr, &state->child_pid, &state->io);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "fork_child failed.\n");
+            goto fail;
+        }
+
+        /* Setup timeout. If failed, terminate the child process. */
+        ret = activate_child_timeout_handler(req, ev,
+                    dp_opt_get_int(kr->krb5_ctx->opts, KRB5_AUTH_TIMEOUT));
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Unable to setup child timeout "
+                  "[%d]: %s\n", ret, sss_strerror(ret));
+            krb5_child_terminate(state->child_pid);
+            goto fail;
+        }
+    } else {
+        /* Continue talking to an existing child. */
+        io_key = talloc_asprintf(state, "%d", kr->pd->child_pid);
```

Comment:
Does this conflict with other krb5_child callers, i.e. What happens if the prompt is this code executes while a krb5_child process is running and has not yet exited (like KCM renewals, or slow krb5 provider child authentication) ?

Reply:
This should not conflict with anything. If pd->child_pid == 0 (not set) a new krb5 child is created.